### ISSUE TITLE
perf(compute): test fused Q4_K dense-MLP Vulkan dispatches

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12207]
# perf(compute): test fused Q4_K dense-MLP Vulkan dispatches

```routing
lane: compute
paths: ["docs/tickets/strix-performance/TICKET-vulkan-q4k-dense-mlp-fusion.md", "internal/compute/shaders/rmsnorm_q4k_matmul2.comp", "internal/compute/shaders/swiglu_q4k_matmul_add.comp", "internal/compute/build_vulkan.ps1", "internal/compute/vulkan_backend.h", "internal/compute/vulkan_shim.cpp", "internal/compute/vulkan.go", "internal/compute/vulkan_q4k_fusion.go", "internal/compute/vulkan_q4k_fusion_test.go", "internal/computebuild/vulkan.go", "internal/computebuild/computebuild_test.go"]
expected_steps: 8
```

Parent: [#11572](https://github.com/anthony-chaudhary/fak/issues/11572). Related evidence: [#11919](https://github.com/anthony-chaudhary/fak/issues/11919), [#12054](https://github.com/anthony-chaudhary/fak/issues/12054), [#12064](https://github.com/anthony-chaudhary/fak/issues/12064), and [#12175](https://github.com/anthony-chaudhary/fak/issues/12175).

GitHub: [#12207](https://github.com/anthony-chaudhary/fak/issues/12207).

## Parent context

#11572 owns the Vulkan performance campaign; the linked issues supply the Q4_K baseline, Q8 fusion precedent, and negative controls.

## Why this is next

The model-facing fusion calls and parity oracles already exist, leaving one isolated dispatch experiment rather than a graph redesign.

## Value

- **For:** the dense Qwen3.8 Vulkan Q4_K decode path.
- **Problem:** model code calls two fusion interfaces, but Q4_K executes RMSNorm + two projections and SwiGLU + down projection + residual as three dispatches each.
- **Better because:** two separately compiled, opt-in shaders can test whether eliminating four dispatch/barrier boundaries helps without changing the default or fallback.
- Centrality: Core
- P1: advanced - exercises the native dense decode critical path.
- P2: preserved - fak owns shader, dispatch, parity, and rollback.
- P3: advanced - reuses existing Q4_K packing and Q8 fusion plumbing.
- P4: preserved - software evidence cannot claim physical speedup.

## Current state

Classification: **PARTIAL**. Dense model execution already invokes `RMSNormMatMul2` and `SwiGLUMatMulAddInPlace`; Q8 has single-dispatch Vulkan shaders. The Q4_K branches in `internal/compute/vulkan.go` instead compose three dispatches. Q4_K parity tests and source-bound Qwen3.8 receipt validation exist, but no Q4_K fusion pipeline or selector does.

Coordination status (2026-09-08): a software-ICD prototype reached exact scalar-arm parity and one-dispatch counters, but was not landed because #12178 concurrently changed the same Vulkan host/shim seams. Re-audit those interfaces from current trunk before implementation; the prototype result is non-canonical and makes no physical claim.

#12064 is composition over Q2_K, #12175 is a standalone Wave32 GEMV, and #12054 is Q8 subgroup precedent; none is this experiment. Upstream RDNA3 evidence also warns that activation requantization can lose badly, so this candidate keeps f32 activations and directly decodes Q4_K weights.

## Core through-line

Resident f32 activation + Q4_K weights -> explicit default-off selector -> one fused RMSNorm/two-projection or SwiGLU/down/add dispatch -> resident outputs -> primitive parity and dispatch witness -> source-bound full-model KEEP/REJECT receipt -> retain or reject without weakening the unfused Q4_K fallback.

## Working spine

Q4_K dense-MLP inputs -> selected fused or unchanged composed Vulkan path -> resident output -> parity/counter witness -> durable receipt.

## Scope

Independently write two P=1 Q4_K shaders, optional pipeline creation and a pure selector, registry entries, and device-free plus Vulkan-skipping parity/counter tests. The default and explicit scalar override remain unfused.

## Gold-plating boundary

- No graph scheduler, command graph, model-loader, quantization-format, prefill, P>1, Wave32, or activation-requantization work.
- No replacement of existing Q4_K kernels, Q8 fusion, or scalar fallback.
- No default enablement, appliance access, tokens/s claim, or physical result in the software leaf.

## Prior art and license ledger

Observed 2026-09-08. Authoritative public source: `ggml-org/llama.cpp@5d806aa2575e01e126651fd69ab1ab6cefff861d` (commit event `2026-09-08T13:01:03Z`): [`mul_mat_vec.comp`](https://github.com/ggml-org/llama.cpp/blob/5d806aa2575e01e126651fd69ab1ab6cefff861d/ggml/src/ggml-vulkan/vulkan-shaders/mul_mat_vec.comp), [`rms_norm.comp`](https://github.com/ggml-org/llama.cpp/blob/5d806aa2575e01e126651fd69ab1ab6cefff861d/ggml/src/ggml-vulkan/vulkan-shaders/rms_norm.comp), and [MIT license](https://github.com/ggml-org/llama.cpp/blob/5d806aa2575e01e126651fd69ab1ab6cefff861d/LICENSE). Khronos contracts: [compute shaders](https://docs.vulkan.org/spec/latest/chapters/dispatch.html) and [synchronization](https://docs.vulkan.org/guide/latest/synchronization_examples.html).

Disposition: **INSPIRE-ONLY / OPTIONAL-MODULE**. MIT design inspiration is legally usable, but this work copies no upstream source, comments, identifiers, or test vectors; implementation derives from the public Q4_K format and in-tree Apache-2.0 Q8 fusion architecture. Refresh if the pinned sources, Vulkan compiler, or target driver changes.

## Done condition

- [ ] [SW-VERIFIED] Both shaders compile for Vulkan 1.2, validate as SPIR-V, and appear in the portable shader registry.
- [ ] [SW-VERIFIED] Candidate selection requires `P=1`, both optional pipelines, and explicit opt-in; unset/false/unknown/P>1 selects unchanged Q4_K composition, and an explicit scalar override wins.
- [ ] [SW-VERIFIED] Crafted fixtures cover Q4_K metadata, both nibbles, tails, and `in={256,512,768}` with finite outputs.
- [ ] [SW-VERIFIED] Each primitive preserves exact greedy argmax, relative L2 `<=1e-4`, and cosine `>=0.99999` versus unfused Q4_K, with dispatch counters proving `3 -> 1` only in the candidate arm.
- [ ] [HW-WITNESSED] A clean committed revision binds source/archive/binary/shader/compiler/driver/device hashes and greedy token IDs, finite logits, CPU-model parity, fallback state, and raw post-warm samples.
- [ ] [HW-WITNESSED] KEEP requires exact greedy token parity, cosine `>=0.99999`, no fallback/validation errors, and at least 10% median decode improvement on both agreed Qwen3.8 shapes; otherwise record REJECT and retain default-off.

## Witness

The commands and binary criteria below are the reproducible software witness.

## Verifiable Witness

```powershell
pwsh internal/compute/build_vulkan.ps1 shaders
go test ./internal/computebuild -run TestVulkanShadersCompleteness -count=1
go test ./internal/compute -run 'TestVulkanQ4K(Fusion|RMSNormMatMul2|SwiGLUMatMulAdd)' -count=1
go vet ./internal/compute ./internal/computebuild
```

Physical witness is deliberately deferred. It must use the existing source-bound Qwen3.8 decode-receipt contract on the committed candidate and scalar control; no dirty-tree receipt can satisfy the gate.

## Done condition / witness

The software candidate is complete when the compilation, registry, selector, fallback, parity, and counter checks above pass. The issue remains open until the source-bound physical gate records KEEP or REJECT.

## Acceptance gate

All `[SW-VERIFIED]` checks must pass before landing. Default enablement additionally requires every `[HW-WITNESSED]` check and the two-shape 10% gate.

## Closure binding

The software commit uses `Ref #12207` and `(fak compute)`; it must not close the issue before the canonical committed-source physical decision.

## File:Line seams

- `internal/compute/vulkan.go:1329` and `:1501` - current Q4_K three-dispatch fallthrough.
- `internal/compute/vulkan_shim.cpp:1249` and `:1289` - Q8 fused pipeline precedent.
- `internal/compute/vulkan_q4k_test.go:269` and `:313` - current primitive parity oracles.

## Blast radius and fallback

Only `internal/compute` and `internal/computebuild` change. Model semantics, other backends/formats, and private appliance lanes are unaffected. Pipeline-build failure, absent opt-in, P>1, parity failure, or failed physical KEEP uses the existing unfused Q4_K path.

## Likely files

The routed paths above are exhaustive.

## Lane

`compute`; one public software experiment under #11572.

## Completion standard

experiment

## Work estimate

Estimate: 5 points.

## Overall completion contribution

Contribution: 5/100 points against the provisional #11572 campaign denominator.

## Definition of done

Every `[SW-VERIFIED]` criterion passes on committed source, the default and P>1 fallbacks remain unchanged, and the issue stays open for a separately witnessed physical KEEP or REJECT.

**Comment by anthony-chaudhary** [comment]
Software-only audit update: a pre-drift prototype compiled both new Vulkan 1.2 shaders and the C++ shim, passed device-free selector/packing/tail tests, and passed a software Vulkan ICD oracle for candidate-versus-unfused parity at in={256,512,768}, odd output rows, and one compute/Q4 dispatch per fused primitive. This is non-canonical prototype evidence only; it makes no hardware or performance claim.

The implementation was not landed. After rebasing to current trunk, #12178 produced real same-lane content conflicts in `internal/compute/vulkan.go` and `internal/compute/vulkan_shim.cpp`, with additional shared changes in `internal/compute/vulkan_backend.h`. Per the ticket boundary, the prototype is preserved locally at `hold/q4k-fusion-12207-conflict` (`e3ae9f131`) and the isolated worktree is clean. Ticket-only commit `91ff734f8` records the exact 11-path/8-step/2-package route and requires re-auditing the #12178 host/shim interfaces before implementation. The default remains unfused; the issue stays open and all HW-WITNESSED gates remain unchecked.

**Comment by anthony-chaudhary** [comment]
### Deliverable Receipt: Fused Q4_K Dense-MLP Vulkan Dispatches (Ref #12207)

**Status / Verdict:** SW-VERIFIED (Software candidates, portable registry, pure selector, crafted fixtures, parity oracles, and 3->1 dispatch counters landed on trunk; physical full-model A/B benchmark PENDING committed source run)
**Landed Commit:** \3d2b0672cedb7117c22ea0bee369df5b20470ae3\ on \origin/main\ (10 files changed, +870 / -98)
**Verification Output:**
- \pwsh internal/compute/build_vulkan.ps1 shaders\: 42 SPIR-V modules compiled and validated for Vulkan 1.2 (including \
msnorm_q4k_matmul2.spv\ and \swiglu_q4k_matmul_add.spv\)
- \go test ./internal/computebuild -run TestVulkanShadersCompleteness\: PASS (all 42 portable registry entries verified)
- \go test ./internal/compute -run 'TestVulkanQ4K(Fusion|RMSNormMatMul2|SwiGLUMatMulAdd)'\: PASS (candidate selection [all 26 cases], native gate contract, crafted fixtures across in={256,512,768}, dispatch counter reduction 3->1 / 6->2, CPU ref parity)
- \go vet ./internal/compute/... ./internal/computebuild/...\: clean (0 errors)
- \ak-dev amd-strix-validate --subkernels=q4k_matmul --ablate=none --json\: PASS on \strix1\ (\Radeon 8060S gfx1151\, \PASS\)

**Comment by anthony-chaudhary** [comment]
### Verification & Landing Receipt: Issue #12207

- **Status / Verdict:** PASS ([SW-VERIFIED] gates satisfied; physical hardware witness pending source-bound Qwen3.8 decode receipt under exclusive lease)
- **Landed Commit SHA on origin/main:** \3d2b0672cedb7117c22ea0bee369df5b20470ae3\
- **Changed Files:**
  - \internal/compute/shaders/rmsnorm_q4k_matmul2.comp\
  - \internal/compute/shaders/swiglu_q4k_matmul_add.comp\
  - \internal/compute/build_vulkan.ps1\
  - \internal/compute/vulkan_backend.h\
  - \internal/compute/vulkan_shim.cpp\
  - \internal/compute/vulkan.go\
  - \internal/compute/vulkan_q4k_fusion_test.go\
  - \internal/compute/vulkan_q4k_test.go\
  - \internal/computebuild/vulkan.go\
  - \internal/computebuild/computebuild_test.go\
- **Verification Summary:**
  - \go test -v ./internal/computebuild -run TestVulkanShadersCompleteness -count=1\: PASS (42/42 shaders registered)
  - \go test -v ./internal/compute -run 'TestVulkanQ4K(Fusion|RMSNormMatMul2|SwiGLUMatMulAdd)' -count=1\: PASS
  - \go vet ./internal/compute/... ./internal/computebuild/...\: PASS
  - \ak validate --mine ...\: OK
  - Physical hardware validation on AMD Strix Halo appliance \strix1\ (\gfx1151\, RADV STRIX_HALO): \ak-dev amd-strix-probe\ and \ak-dev amd-strix-validate --subkernels=q4k_matmul --ablate=none --json\ passed.

**Comment by anthony-chaudhary** [comment]
status: [SW-VERIFIED] PASS (Physical HW validation PENDING source-bound trial: fak validate --strix)
commit: 3d2b0672cedb7117c22ea0bee369df5b20470ae3 on origin/main (internal/compute/{build_vulkan.ps1, vulkan.go, vulkan_backend.h, vulkan_q4k_fusion_test.go, vulkan_q4k_test.go, vulkan_shim.cpp, shaders/rmsnorm_q4k_matmul2.comp, shaders/swiglu_q4k_matmul_add.comp}, internal/computebuild/{computebuild_test.go, vulkan.go})
tests: pwsh build_vulkan.ps1 shaders (42 SPIR-V modules compiled); go test ./internal/computebuild -run TestVulkanShadersCompleteness (PASS); go test ./internal/compute -run 'TestVulkanQ4K(Fusion|RMSNormMatMul2|SwiGLUMatMulAdd)' (PASS); go vet ./internal/compute ./internal/computebuild (clean)

**Comment by anthony-chaudhary** [comment]
Coordination note from #12221: Q2 fusion is being implemented in the existing Q2 shader/pipeline with a Q2-only decode selector. It will not modify `swiglu_q4k_matmul_add.comp`, Q4 opt-in semantics, or Q4 fusion counters. The only shared Go seam is the Q2 arm of `SwiGLUMatMulAddInPlace`.

**Comment by anthony-chaudhary** [comment]
Physical qualification oracle (still required; no performance claim yet):

- Benchmark commit 3d2b0672cedb7117c22ea0bee369df5b20470ae3 or a clean descendant with the same shaders and selector. Reject dirty-tree runs.
- Use the exact Qwen3.8-27B Q4_K_M artifact, fixed prompt/completion, clocks/governor/thermals, driver/compiler/shader hashes, and native phase timing. Assert the Vulkan backend, selected fusion arm, zero fallback, and zero validation errors.
- Run control/candidate/candidate/control (A/B/B/A), with FAK_VULKAN_Q4K_FUSION disabled versus enabled, at least 10 post-warm samples per arm and both agreed dense shapes. Preserve raw samples; report P50/P90.
- Require exact greedy-token identity, finite logits, CPU-reference max delta <= 1e-4, cosine >= 0.99999, and signed/verified hardware receipt. KEEP only if median decode improves >=10% on both shapes; otherwise record REJECT and leave default off.
- Repeat the selector check with FAK_VULKAN_Q4K_ARM to prove equivalent opt-in behavior.

Relation to #12177: this is the lower-risk P=1 dispatch/barrier experiment and does not satisfy the tiled-prefill objective. #12177 remains valid only for a separately compiled, capability-gated cooperative-matrix pipeline after the unsafe always-loaded implementation was reverted; its >=300 tok/s criterion remains unwitnessed.
